### ISSUE TITLE
[Suggestion] `/clayout`

### ISSUE THREAD

**Issue by barraIhsan** [Earthcomputer/clientcommands#326]
## Requirement

-   Not in spectator mode

## Syntax

`/clayout (create|edit) block <x> <y> <z> <item_slot> <name>`
`/clayout (create|edit) entity <entity> <item_slot> <name>`
`/clayout (delete|preview) <name>`
`/clayout load <name> [--drop-new-item]`
`/clayout list`

### `item_slot`

Must be a string notation that refers to certain slots in the inventory.
| Slot | Valid *slot_child* && *slot_number* |
| --- | --- |
| armor | |
| armor._slot_child_ | `head`, `chest`, `legs`, `feet` |
| weapon | |
| weapon._slot_child_ | `mainhand`, `offhand` |
| container | |
| container._slot_child_ | 0 - 53 |
| enderchest | |
| enderchest._slot_child_ | |
| hotbar | | |
| hotbar._slot_child_ | |
| inventory | | |
| inventory._slot_child_ | |
| horse | |
| horse._slot_child_ | `saddle`, `chest`, `armor`, 0 - 14 |
| ~~villager~~ | |
| ~~vilalger._slot_child_~~ | ~~0 - 7~~ |

### `name`

Optional. The name of the layout

## Description
`create` : Create a new layout from the items in `item_slot`
`edit` : Edits a layout, replacing items in the layout with the items in `item_slot`
`delete` : Deletes an existing layout.
`preview` : Displays a `item_slot` with the items contained in the layout.
`load ` : Load a layout, new item (an item that not in the layout) will be ignored
`load --drop-new-item` : Load a layout, new item (an item that not in the layout) will be dropped

## Data
The data will be stored at `.minecraft\config\clientcommands\layout.dat` (Its also has the `dat_old` version for backup). And the structure looks like this (SNBT Format, Just like `group.dat` and `kits.dat`) :
```
{
	"DataVersion": <data_version>,
	"Layout": {
		"<name>": [
			{
				"source": "<item_slot>",
				"Slot": <slot_number>,
				"id": "<namespace>:<item_id>",
				"Count": <item_count>b,
				"tag": {
					<item_tag> (if exist)
				}
			}
		]
	}
}
```

What is the `source` for? It's for displaying the preview

## Preview Layout
<details><summary>Click to see the preview layout</summary>

* [`armor`](https://user-images.githubusercontent.com/57800056/137657614-65ff4980-bf11-4695-920b-dee4880b14fe.png)
	* [`armor.head`](https://user-images.githubusercontent.com/57800056/137657877-6d5fd6d1-3789-493e-8edc-242b43d5f698.png)
	* [`armor.chest`](https://user-images.githubusercontent.com/57800056/137657915-615659c9-e981-4cec-9434-f912d8b33e98.png)
	* [`armor.legs`](https://user-images.githubusercontent.com/57800056/137658003-e5559bbb-1420-4a71-ad15-a893644b28a0.png)
	* [`armor.feet`](https://user-images.githubusercontent.com/57800056/137658037-7beb7431-8621-4c13-b753-05dfae83ba65.png)
* [`weapon`](https://user-images.githubusercontent.com/57800056/137658539-9ade9dcf-c4ab-4139-8f98-ab75c0b7d803.png)
	* [`weapon.mainhand`](https://user-images.githubusercontent.com/57800056/137658595-f942d649-bd0a-492d-9d7a-834d67a96806.png)
	* [`weapon.offhand`](https://user-images.githubusercontent.com/57800056/137658622-c89ecf6f-f47a-4cdf-b789-9346d4da58bf.png)
* [`container` (`barrel`, `chest`, `chest_minecart`, `shulker_box`, `trapped_chest`)](https://user-images.githubusercontent.com/57800056/137708487-3b50a0cd-4f02-4242-8775-cb2bcae4fb75.png) ([`large_chest`, and `large_trapped_chest`](https://user-images.githubusercontent.com/57800056/137710118-c39ea0ed-ae92-4d5a-bc81-bf06c22b74d1.png), [`furnace`, `blast_furnace`, and `smoker`](https://user-images.githubusercontent.com/57800056/137710686-a71a9a87-7fbf-4f27-9c3b-5d5eefd21e48.png), [`dispenser`, and `dropper`](https://user-images.githubusercontent.com/57800056/137710874-dfc21638-11b8-4615-8451-a0f18c5aed6f.png), [`brewing_stand`](https://user-images.githubusercontent.com/57800056/137711183-4a3f5584-0850-46b7-a737-79f5b6807604.png), [`hopper`, and `minecart_hopper`](https://user-images.githubusercontent.com/57800056/137711974-9b1fca29-ecf0-4832-88f4-f379ded2bec6.png))
	* [`container.slot_number`](https://user-images.githubusercontent.com/57800056/137658595-f942d649-bd0a-492d-9d7a-834d67a96806.png)
* [`enderchest`](https://user-images.githubusercontent.com/57800056/137659753-0c054721-ae76-4b55-b036-04304c02afb2.png)
	* [`enderchest.slot_number`](https://user-images.githubusercontent.com/57800056/137658595-f942d649-bd0a-492d-9d7a-834d67a96806.png)
* [`hotbar`](https://user-images.githubusercontent.com/57800056/137659039-ad2ef405-0868-4bca-b1b2-551aaf8854b4.png)
	* [`hotbar.slot_number`](https://user-images.githubusercontent.com/57800056/137658595-f942d649-bd0a-492d-9d7a-834d67a96806.png)
* [`inventory`](https://user-images.githubusercontent.com/57800056/137708487-3b50a0cd-4f02-4242-8775-cb2bcae4fb75.png)
	* [`inventory.slot_number`](https://user-images.githubusercontent.com/57800056/137658595-f942d649-bd0a-492d-9d7a-834d67a96806.png)
* [`horse`](https://user-images.githubusercontent.com/57800056/137662029-5ade2429-922e-4825-9dc8-efa9e49c754d.png) ([`llamas w/ chest`](https://user-images.githubusercontent.com/57800056/137662053-c61ba56c-c43b-41ba-a562-7f4521d558f7.png), [`llamas w/o chest`](https://user-images.githubusercontent.com/57800056/137662107-f6de34c6-0267-40f2-8c0a-22ea4b81473c.png), [`donkey w/ chest`, and `mule w/ chest`](https://user-images.githubusercontent.com/57800056/137662255-2bfcc454-b52a-4d7d-9db0-165c5aaccc68.png), [`donkey w/o chest`, `mule w/o chest`](https://user-images.githubusercontent.com/57800056/137662209-e87976ac-dc2e-4dcd-a9b1-96a72047f519.png))
	* [`horse.saddle`](https://user-images.githubusercontent.com/57800056/137662397-099de197-ffd1-482c-a347-0151ec9dbfd4.png)
	* [`horse.chest`](https://user-images.githubusercontent.com/57800056/137662441-0b55c1fc-c92f-46d2-9146-ffda87119f04.png)
	* [`horse.armor`](https://user-images.githubusercontent.com/57800056/137662476-699a26da-a8a6-49b2-95f3-14372532fafb.png)

</details>

## Temp Rule
There is also a new temp rule called `createLayoutBeforeDeath`. It will create a new layout
(If index 0 exist, then the index is 1, etc.)
`death_armor_<index>`
`death_weapon_<index>`
`death_inventory_<index>`
`death_hotbar_<index>`
before you die. So you can load them after you pick up all the items to organize back your inventory.

## Usage + Video Example

You can create a layout and then load them.
For example, you can create the layout before death. And after picking up all the items, you can load the layout to organize back your inventory.

Video Example
(If you see this, I not yet record an updated video, it should be `/clayout create entity @p inventory test` instead of `/clayout create @p inventory test`

https://user-images.githubusercontent.com/57800056/137707651-b2010d78-0166-4ae8-bae8-d6580c4f2a85.mp4
 
Note: If you load a layout, it will take an item from your **armor, weapon, inventory and hotbar**

**Comment by xpple** [comment]
So this is similar to the `ckit` command but it takes the items out of your inventory and rearranges them?

**Comment by barraIhsan** [comment]
yes, it is gonna useful for survival. So you don't have to **layout** your inventory back
